linked_list/linked_list.py

```diff
@@ -1,79 +1,166 @@
-        pass
-
+        if self.head:
+            return self.head.value
+        return None
 
+    #----------------------
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
-
+        if self.head is None:
+            self.head = self.tail = Node(value)
+        else:    
+            new_node = Node(value, self.head)
+            self.head.previous = new_node
+            self.head = new_node
+
+    #----------------------
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current = self.head
+        if current == None:
+            return False
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
+    #----------------------
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter = 0
+        current = self.head
+
+        while current != None:
+            current = current.next
+            counter +=1
+        return counter
 
+    #----------------------
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        counter = 0
+        current = self.head
+
+        while current != None:
+            if counter == index:
+                return current.value
+            counter += 1 
+            current = current.next
+        return None
 
+    #----------------------
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+        if current == None:
+            return None
+
+        while current.next != None:
+            current = current.next
+        return current.value
 
+    #----------------------
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):

```

Comment:
✨ However, again, since you have a `tail` node, you could just make the new node the new tail and redirect the pointers of the current tail for an O(1) time complexity solution